File: python/datagen/datagen.py
from random import gauss
from datetime import datetime, timedelta
from functools import partial
import operator
import math
import sys
import os

def seq_gen(start, op=partial(operator.add, 1)):
  while True:
    yield start
    start = op(start)

def fn_gen(fn):
  while True:
    yield fn()

class cyclic_moving_body:
  def __init__(self):
    self.weight = 0.1
    self.time = datetime.utcnow()
    self.pos = 0
    self.speed = 0
    self.acc = 0

  def next(self):
    dt = timedelta(milliseconds=gauss(500, 300))
    self.time += dt  # update time
    dv = self.acc * dt.total_seconds()  # dv = a * t
    dp = self.speed + dv * dt.total_seconds() / 2 # dp = v + (a * t^2)/2 = v + dv * t / 2
    self.speed += dv  # update speed
    self.pos += dp  # update position

    acc_bias = 0.05 * math.sin(self.time.second / 10 * 2 * math.pi)  # interval is 10 seconds
    self.acc = gauss(0, 0.01) + acc_bias
    self.acc = math.copysign(self.acc, -self.pos)  # acc always try to bring pos to 0

    return '{0.time},{0.acc},{0.speed},{0.pos}'.format(self)

if __name__ == "__main__":
  if(len(sys.argv) < 3):
    print("Usage: {} <out_file> <approx_size>").format(os.path.basename(__file__))
  else:
    out_file = sys.argv[1]
    out_min_bytes = int(sys.argv[2])

    with open(out_file, 'w') as fd:
      seq_number = seq_gen(0)
      time = seq_gen(datetime.utcnow(), lambda v: v + timedelta(milliseconds=gauss(500, 300)))
      weight = fn_gen(partial(gauss, 12, 3))
      body_info = cyclic_moving_body()

      generators = [seq_number, time, weight, body_info]
      while fd.tell() < out_min_bytes:
        line = ''
        for g in generators:
          line += str(g.next()) + ','
        fd.write(line[:-1] + '\n')
        # print line[:-1]

File: scala/dataGen/src/dataGen.scala
import java.io._
import breeze.stats.distributions.Gaussian

class cyclicMovingBody() {
  private
  def move(): Unit ={
    val dt = System.currentTimeMillis() - time
    time += dt
    val dv = acc * dt  // dv = a * t
    val dp = speed + dv * dt / 2 // dp = v + (a * t^2)/2 = v + dv * t / 2
    speed += dv // update speed
    pos += dp   // update position

    val acc_bias = 0.05 * math.sin(time / 10000 * 2 * math.Pi)  // interval is 10 seconds
    acc = Gaussian(0, 0.01).draw() + acc_bias
    acc = Math.copySign(acc, -pos)  // acc always try to bring pos to 0
  }

  override def toString() : String = acc.toString + "," + speed.toString + "," + pos.toString

  def getCurrentState(): String = {
    move()
    return toString()
  }

  private var time: Long = System.currentTimeMillis()
  private var pos: Double = 0
  private var speed: Double = 0
  private var acc: Double = 0
}

object dataGen {
  def seqGen[T](start: T, op: T => T): Stream[T] = start #:: seqGen(op(start), op)

  def main (args: Array[String]): Unit = {
    if(args.length < 2){
      println("Usage: dataGen <out_file> <approx_size>")
    }
    else {
      val outFile = args(0)
      val outMinBytes = args(1).toLong

      val seqNumber = Stream.iterate(0)(_ + 1).iterator
      val time = Stream.iterate(System.currentTimeMillis())(_ + Gaussian(500, 300).draw().asInstanceOf[Long]).iterator
      val weight = Stream.continually(Gaussian(13, 3).draw()).iterator
      val body = new cyclicMovingBody
      val bodyData = Stream.continually(body.getCurrentState()).iterator

      val iterators = List(seqNumber, time, weight, bodyData)

      val bw = new BufferedWriter(new FileWriter(outFile))
      var totalWritten: Long = 0

      val startTime = System.currentTimeMillis()
      while(totalWritten < outMinBytes)
      {
        val str = iterators.map(_.next()).mkString(",") + "\n"
        bw.write(str)
        totalWritten += str.length
      }
      bw.close()
      val totalTime_s = (System.currentTimeMillis() - startTime) / 1000
      println(s"Finished in ${totalTime_s} seconds, writing ${totalWritten / 1024/1024 / totalTime_s} Mb/s")
    }
  }
}
